mlflow/types/utils.py

```diff
@@ -78,6 +87,117 @@ def _get_str_or_byte_type(data):
         return DataType.binary
 
 
+def _infer_datatype(data) -> DataType:
+    if DataType.is_boolean(data):
+        return DataType.boolean
+    if DataType.is_integer(data):
+        return DataType.integer
+    if DataType.is_long(data):
+        return DataType.long
+    if DataType.is_float(data):
+        return DataType.float
+    if DataType.is_double(data):
+        return DataType.double
+    if DataType.is_string(data):
+        return DataType.string
+    if DataType.is_binary(data):
+        return DataType.binary
+    if DataType.is_datetime(data):
+        return DataType.datetime
+
+
+def _update_object_properties(obj1: Object, obj2: Object) -> Object:
+    if obj1 == obj2:
+        return obj1
+    properties1 = {prop.name: prop for prop in obj1.properties}
+    properties2 = {prop.name: prop for prop in obj2.properties}
+    updated_properties = []
+    # For each property in the first element, if it doesn't appear
+    # later, we update required=False
```

Comment:
why do we need this `required` update logic?

Reply:
Because for List[Dict], it's possible that first dictionary has some property, but the second dictionary doesn't have that property, then we should update that property required=False instead of saying they're not compatible.
```
[{"messages": ..., "name": ...}, {"messages": ...}] 
```
then `name` property should be set as required=False. And those two objects in the array still share the same schema.